Create templates for documenting triggers, conditions & actions (#3065)
home-assistant/developers.home-assistant: docs/documenting/integration-docs-examples.md

This page outlines the high-level structure of an integration page.
Use this structure together with the following documentation:

- [Documentation standards](/docs/documenting/standards).
- [Documentation style guide](/docs/documenting/general-style-guide).
- The documentation rules of the [Integration Quality Scale](/docs/core/integration-quality-scale/).

## Basic structure of an integration page

> For a template that you can copy and paste, see the integration documentation template in the `home-assistant.io` repository: [/_integrations/_integration_docs_template.markdown](https://github.com/home-assistant/home-assistant.io/tree/current/source/_integrations/_integration_docs_template.markdown).

Author: ℹ️ Moved this here to make sure nobody misses it while going directly to this section 😅

- Configuration options
- Supported functionality
- Actions
- Conditions

> - Triggers

Reviewer: ```suggestion
- List of actions
- List of conditions
- List of triggers
```
If these no longer contain the actual description but instead provide the full list of all available items, I would indicate that in the title.

<img width="877" height="318" alt="Image" src="https://github.com/user-attachments/assets/891a111a-6b4d-4959-ae3f-4bf55854f12d" />

Author: @c0ffeeca7 alright, let's do it.

Author: I'll make a PR for the includes `integrations/actions`, `integrations/conditions` and `integrations/triggers`.

- Examples
- Data updates
- Known limitations
- Troubleshooting
- Community notes
- Removing the integration

## Documenting automation triggers, conditions, and actions

When you document an integration, create a separate file for each of its triggers, conditions, and actions.
Then include them in the main integration page with:

```liquid
{% include integrations/triggers_conditions_actions.md %}
```

### Template: trigger

Create a file in the `home-assistant.io` repository in `source/_triggers`.
Save it as `<my_integration>.<trigger_name>.markdown`, for example: `light.brightness_changed.markdown`.

Adapt this template for your integration:

```md
---
title: "Light brightness changed"
trigger: light.brightness_changed
domain: light
description: "Triggers after the brightness of one or more lights changes."
related_triggers:
  - light.brightness_crossed_threshold
  - light.turned_on
---

The **Light brightness changed** trigger fires after...
Use it to...

{% include integrations/labs_entity_triggers_note.md %}

{% include triggers/ui_header.md %}

> To use this trigger in an automation:

Reviewer: ```suggestion
## Using this trigger in an automation
```
Should this be a title? It introduces a task topic.

Author: @c0ffeeca7 Yep, good point!

Author: @c0ffeeca7 Actually, no need, because the include above inserts a h2:

<img width="785" height="339" alt="image" src="https://github.com/user-attachments/assets/8d847236-a57a-4a65-b063-c2d75639d166" />

1. Go to {% my automations title="**Settings** > **Automations & scenes**" %}.
2. Open an existing automation, or select **Create automation** > **Create new automation**.
3. In the **When** section, select **Add trigger**.
4. From the search box, search for and select **Light: Light brightness changed**.
5. Select what you want to monitor. Under **By target** (see [Targets](#targets)), pick... You can also select a floor, a device, a specific entity, or a label.
6. From the triggers shown for that target, select...
7. Under **Trigger when** (see [Behavior](#behavior-with-multiple-targets)), pick **Any**, **First**, or **Last** to control how multiple targets interact.
8. Under **Another field**, set...

> 9. Select **Save**.

Author: Add Treshold example:

```suggestion
7. Under **Trigger when** (see [Behavior](#behavior-with-multiple-targets)), pick **Any**, **First**, or **Last** to control how multiple targets interact.
8. Under **Threshold type**, set how much the level has to change before the trigger fires.
9. Under **Another field**, set...
10. Select **Save**.

### Options in the UI

{% options_ui %}
Threshold type:
  description: How much the brightness has to change before the trigger fires, as a percentage of full brightness. Can be a fixed number, or reference a helper entity that provides the value.
  required: true
{% endoptions_ui %}

{% include triggers/yaml_header.md %}

In YAML, refer to this trigger as `light.brightness_changed`. A basic example looks like this:

{% example %}
trigger: |
  trigger: light.brightness_changed
  target:
    entity_id: light.living_room
  options:

>     threshold: 10

Author: Add example with a default value:

```suggestion
    threshold: 10
behavior:
  description: >
    When multiple sensors are targeted, controls when the trigger fires. Accepts `any`, `first`, or `last`.
  required: true
  type: string
  default: any
```

{% endexample %}

This fires whenever the living room light's brightness changes by at least ten percent.

### Options in YAML

YAML sometimes provides additional options for more complex use cases that are not available through the UI.

{% options_yaml %}
threshold:
  description: >
    The minimum amount (in percent) the brightness must change before the trigger fires. Accepts a number, or a reference to an `input_number`, `number`, or `sensor` entity with a percent unit.
  required: true
  type: any
{% endoptions_yaml %}

<!-- Keep the "include" below if your integration supports targets -->
{% include triggers/targets.md %}

> {% include triggers/behavior.md %}

Author: This depends on https://github.com/home-assistant/home-assistant.io/pull/44955 for the new include.

- **Action**: Fan: Set speed

{% details "YAML example for a ceiling-light-linked fan" %}

{% example %}
automation: |
  alias: "Match fan to ceiling light"
  triggers:
    - trigger: light.brightness_changed
      target:
        entity_id: light.living_room_ceiling
      options:
        threshold: 10
  actions:
    - action: fan.set_percentage
      target:
        entity_id: fan.living_room
      data:
        percentage: "{{ state_attr('light.living_room_ceiling', 'brightness_pct') | int }}"
{% endexample %}

{% enddetails %}

{% include triggers/stuck.md %}

{% include triggers/related.md %}
```

### Template: condition

Create a file in the `home-assistant.io` repository in `source/_conditions`.
Save it as `<my_integration>.<condition_name>.markdown`, for example: `light.is_on.markdown`.

Adapt this template for your integration:

```md
---
title: "Light is on"
condition: light.is_on
domain: light
description: "Tests if one or more lights are on."
related_conditions:
  - light.is_off
  - light.is_brightness
---

The **Light is on** condition passes when...
Use it to...

{% include integrations/labs_entity_triggers_note.md %}

{% include conditions/ui_header.md %}

> To use this condition in an automation:

Reviewer: ```suggestion
## Using this condition in an automation
```

  conditions:
    - condition: light.is_on
      target:
        entity_id: light.living_room
      options:
        behavior: any
  actions:
    - action: media_player.play_media
      target:
        entity_id: media_player.living_room
      data:
        media_content_id: "media-source://tts/cloud?message=Someone+is+at+the+door"
        media_content_type: music
        announce: true
{% endexample %}

{% enddetails %}

{% include conditions/stuck.md %}

{% include conditions/related.md %}

```

### Template: action

Create a file in the `home-assistant.io` repository in `source/_actions`.
Save it as `<my_integration>.<action_name>.markdown`, for example: `light.turn_on.markdown`.

Adapt this template for your integration:

```md
---
title: "Turn on a light"
action: light.turn_on
domain: light
description: "Turn a light on. Optionally set brightness, color, color temperature, an effect, or a transition."
since: "0.7"
related_actions:
  - light.turn_off
  - light.toggle
---

The **Turn on** action turns a light on...
This action works with any light {% term entity %} in Home Assistant...
If the light is already on, calling the action...

{% include actions/ui_header.md %}

> To turn a light on from an automation or a script:

Reviewer: ```suggestion
## Turning on a light on from an automation or a script
```
